Fix LineEdit and TextEdit double-click and triple-click selection

jmb462 opened the pull request:
Fixes #52975

### Issue description : 

Double clicking after the last word in a LineEdit should select the last word.
Triple clic anywhere (so after the last word too) in the LineEdit should select all content.
These two behavior are broken in 4.0 (regression)

### Cause :

There were 2 differents problems here :

Double clic word selection didn't catch the case where the carret was exactly at the end of the last word.
(We don't want to just replace test ``words[i].y > caret_column`` with ``words[i].y >= caret_column`` here cause we don't want to select the word before a space between two words)

If no word was selected after a double-click (if the last character is not a letter/number or because of the first part of the bug above), the last time of double click wasn't updated and so triple click was never detected. So I moved the last time update to be updated even if no word is selected after a double-clic.

### Before

![before](https://user-images.githubusercontent.com/3649998/134551818-cc9842c8-8c9a-43c8-a674-b2c94bcc4c5b.gif)
(Double click after the last word doesn't select it and last double-clic timer is not updated so triple clic is broken too)

![before2](https://user-images.githubusercontent.com/3649998/134551929-44100deb-746f-4ff4-b2fa-ea06c054c566.gif)
(Double click at the end doesn't select word because last character is not a letter or number AND triple click is never detected)
 
### After (Same behaviour as in3.x) :

![after](https://user-images.githubusercontent.com/3649998/134551837-2233cb42-79d7-4587-9ddb-1b4d49cea1a5.gif)
(Double click after the last word select it and last double-clic timer is updated so triple clic works fine)

![after2](https://user-images.githubusercontent.com/3649998/134551945-a90b790c-fab9-4da9-9c7c-2f5b03ea28f4.gif)
(Double click at the end doesn't select word because last character is not a letter or number BUT triple click is correctly catch)

jmb462 committed:
Fix LineEdit and TextEdit double-click and triple-click selection

Paulb23 commented on scene/gui/line_edit.cpp:
```diff
@@ -270,14 +270,14 @@ void LineEdit::gui_input(const Ref<InputEvent> &p_event) {
 						caret_column = selection.begin;
 					} else if (b->is_double_click()) {
 						// Double-click select word.
+						selection.last_dblclk = OS::get_singleton()->get_ticks_msec();
```
Not sure about moving this out of the loop, if there is no text and you triple click we will create a section of length `0`, which might cause some issues. Either we put this back in the loop, or triple click needs to check the text length.

jmb462 replied:
OK, looking at 3.x code and seems that there was no check of that sort.

But you're right, it may be be safer to ensure that text length is > 0 in the triple click block.

Paulb23 reviewed (approved):
LGTM, not sure if want to do it as part of this PR but, looks like `TextEdit` needs the same fix as well.

jmb462 commented:
> LGTM, not sure if want to do it as part of this PR but, looks like `TextEdit` needs the same fix as well.

I'm looking at it.

Additionnaly, I saw that there is a 5 pixels tolerance for triple click detection in TextEdit.
Shouldn't we add this tolerance to LineEdit too for consistancy ?

Paulb23 commented:
> Shouldn't we add this tolerance to LineEdit too for consistancy ?

Yep, that sounds good.

jmb462 commented:
So, for consistancy reason : 
In LineEdit :
- added a 5 pixels tolerance to triple-click detection (added a member variable ``last_dblclk_pos`` as in TextEdit)
- moved ``last_dblclk`` outside the selection structure (as in TextEdit). Doesn't need to be part of that structure.

In TextEdit : 
- allow to select the last word of the line by double-clicking after the last word of the line. (as in LineEdit and same behavior as VisualStudio / SublimeText)
- check if line is non-empty for triple-click block

Paulb23 reviewed (changes requested):
For `TextEdit` we have to allow triple click on empty lines, to allow users to start to dragging the selection. `LineEdit`, gets away with it as there's only one line. 

Otherwise changes looks good.

jmb462 commented:
> For `TextEdit` we have to allow triple click on empty lines, to allow users to start to dragging the selection. `LineEdit`, gets away with it as there's only one line.
> 
> Otherwise changes looks good.

OK, sorry. Amended.

akien-mga commented:
Thanks!
